### PR TITLE
Bump Golang to 1.10.1

### PR DESCRIPTION
**Pull request by thaJeztah**
Let's see what breaks (or not!)

From the announcement:

Hello gophers,

We just released Go 1.10.

You can read the announcement blog post here:
  https://blog.golang.org/go1.10

You can download binary and source distributions from our download page:
  https://golang.org/dl/

To compile from source using a Git checkout, update to the release with "git checkout go1.10" and build as usual.

To find out what has changed, read the release notes:
  https://golang.org/doc/go1.10

Thanks to everyone who contributed to the release.

Andy on behalf of the Go Team

### CODE REVIEW

File: Dockerfile.e2e [hunk 1]
```diff
@@ -1,5 +1,5 @@
 ## Step 1: Build tests
-FROM golang:1.9.2-alpine3.6 as builder
+FROM golang:1.10beta1-alpine3.6 as builder
```
Comment (thaJeztah) [inline Dockerfile.e2e]: Image is not up yet, so this one won't work (yet)

**Comment by thaJeztah** [comment]
Release notes: https://beta.golang.org/doc/go1.10

**Comment by thaJeztah** [comment]
We can start experimenting with `go test -json` 🎉

**Comment by tophj-ibm** [comment]
FYI power failure is legit and we're looking at it. There is a leaky fd somewhere which is causing a semaphore introduced in 1.10 to not be released

**Comment by cpuguy83** [comment]
On windows:

<details>

```
2:07:05 PANIC: docker_api_logs_test.go:152: DockerSuite.TestLogsAPIUntil
02:07:05 
02:07:05 ... Panic: runtime error: index out of range (PC=0x458491)
02:07:05 
02:07:05 d:/CI/CI-ff7e5d9ca/go/src/runtime/asm_amd64.s:573
02:07:05   in call32
02:07:05 d:/CI/CI-ff7e5d9ca/go/src/runtime/panic.go:502
02:07:05   in gopanic
02:07:05 d:/CI/CI-ff7e5d9ca/go/src/runtime/panic.go:28
02:07:05   in panicindex
02:07:05 docker_api_logs_test.go:175
02:07:05   in DockerSuite.TestLogsAPIUntil
02:07:05 d:/CI/CI-ff7e5d9ca/go/src/runtime/asm_amd64.s:573
02:07:05   in call32
02:07:05 d:/CI/CI-ff7e5d9ca/go/src/reflect/value.go:447
02:07:05   in Value.call
02:07:05 d:/CI/CI-ff7e5d9ca/go/src/reflect/value.go:308
02:07:05   in Value.Call
02:07:05 c:/gopath/src/github.com/docker/docker/vendor/github.com/go-check/check/check.go:816
02:07:05   in suiteRunner.forkTest.func1
02:07:05 c:/gopath/src/github.com/docker/docker/vendor/github.com/go-check/check/check.go:672
02:07:05   in suiteRunner.forkCall.func1
02:07:05 d:/CI/CI-ff7e5d9ca/go/src/runtime/asm_amd64.s:2361
02:07:05   in goexit
02:07:09 
```
</details>

**Comment by cpuguy83** [comment]
That one could be a test issue, but re-running.

**Comment by tonistiigi** [comment]
@thaJeztah You can remove the vendored `archive/tar` in this PR as it should be fixed in `go1.10`

**Comment by thaJeztah** [comment]
@tonistiigi thanks! removed 👍

File: Dockerfile [hunk 2]
```diff
@@ -87,7 +87,7 @@ RUN apt-get update && apt-get install -y \
 #            will need updating, to avoid errors. Ping #docker-maintainers on IRC
 #            with a heads-up.
 # IMPORTANT: When updating this please note that stdlib archive/tar pkg is vendored
```
Comment (tonistiigi) [inline Dockerfile]: remove these `archive/tar` comment lines as well

Comment (thaJeztah) [reply Dockerfile]: Oh! forgot about that one

**Comment by thaJeztah** [comment]
Bunch of failures now due to the Archive changes;

```
20:26:23 --- FAIL: TestDispatch (0.06s)
20:26:23 	evaluator_test.go:127: Error when creating tar context: Error processing tar file(exit status 1): unexpected EOF
20:26:23 FAIL
20:26:23 coverage: 42.6% of statements
20:26:23 FAIL	github.com/docker/docker/builder/dockerfile	0.345s

20:26:29 --- FAIL: TestHashFile (0.07s)
20:26:29 	tarsum_test.go:146: Error when executing FromArchive: Error processing tar file(exit status 1): unexpected EOF
20:26:29 --- FAIL: TestHashSubdir (0.06s)
20:26:29 	tarsum_test.go:146: Error when executing FromArchive: Error processing tar file(exit status 1): unexpected EOF
20:26:29 FAIL
20:26:29 coverage: 13.0% of statements
20:26:29 FAIL	github.com/docker/docker/builder/remotecontext	0.294s

20:27:57 --- FAIL: TestChrootTarUntarWithSymlink (0.05s)
20:27:57 	archive_test.go:194: Unexpected differences after untar: [{/file-1 C} {/file-2 C} {/file-5 C} {/file-6 C} {/file-7 C} {/file-8 C} {/file-9 C} {/file-0 C} {/file-4 C} {/file-3 C}]
20:27:57 FAIL
20:27:57 coverage: 37.0% of statements
20:27:57 FAIL	github.com/docker/docker/pkg/chrootarchive	2.997s


20:28:40 --- FAIL: TestTarSumRemoveNonExistent (0.00s)
20:28:40 	builder_context_test.go:27: failed to read from testdata/46af0962ab5afeb5ce6740d4d91652e69206fc991fd5328c1a94d364ad00e457/layer.tar: archive/tar: missed writing 3348 bytes
20:28:40 --- FAIL: TestTarSumRemove (0.00s)
20:28:40 	builder_context_test.go:57: failed to read from testdata/46af0962ab5afeb5ce6740d4d91652e69206fc991fd5328c1a94d364ad00e457/layer.tar: archive/tar: missed writing 3348 bytes
20:28:40 --- FAIL: TestTarSums (0.04s)
20:28:40 	tarsum_test.go:370: failed to copy from testdata/46af0962ab5afeb5ce6740d4d91652e69206fc991fd5328c1a94d364ad00e457/layer.tar: archive/tar: missed writing 3348 bytes
20:28:40 	tarsum_test.go:370: failed to copy from testdata/46af0962ab5afeb5ce6740d4d91652e69206fc991fd5328c1a94d364ad00e457/layer.tar: archive/tar: missed writing 3348 bytes
20:28:40 	tarsum_test.go:370: failed to copy from testdata/46af0962ab5afeb5ce6740d4d91652e69206fc991fd5328c1a94d364ad00e457/layer.tar: archive/tar: missed writing 3348 bytes
20:28:40 	tarsum_test.go:370: failed to copy from : archive/tar: missed writing 8192 bytes
20:28:40 	tarsum_test.go:363: failed to read 16KB from testdata/collision/collision-0.tar: archive/tar: missed writing 6 bytes
20:28:40 	tarsum_test.go:363: failed to read 16KB from testdata/collision/collision-1.tar: archive/tar: missed writing 6 bytes
20:28:40 	tarsum_test.go:363: failed to read 16KB from testdata/collision/collision-2.tar: archive/tar: missed writing 6 bytes
20:28:40 	tarsum_test.go:363: failed to read 16KB from testdata/collision/collision-3.tar: archive/tar: missed writing 6 bytes
20:28:40 	tarsum_test.go:370: failed to copy from : archive/tar: missed writing 8192 bytes
20:28:40 	tarsum_test.go:370: failed to copy from : archive/tar: missed writing 8192 bytes
20:28:40 	tarsum_test.go:370: failed to copy from : archive/tar: missed writing 8192 bytes
20:28:40 	tarsum_test.go:370: failed to copy from : archive/tar: missed writing 8192 bytes
20:28:40 	tarsum_test.go:370: failed to copy from : archive/tar: missed writing 8192 bytes
20:28:40 --- FAIL: TestIteration (0.00s)
20:28:40 	tarsum_test.go:511: archive/tar: missed writing 4 bytes
20:28:40 FAIL
20:28:40 coverage: 90.4% of statements
20:28:40 FAIL	github.com/docker/docker/pkg/tarsum	0.051s
```

**Comment by thaJeztah** [comment]
Need to verify if this fixes https://github.com/moby/moby/issues/29360

**Comment by thaJeztah** [comment]
Bumped to Go 1.10beta2

**Comment by thaJeztah** [comment]
Bumped to 1.10rc1

**Comment by Mic92** [comment]
go 1.10 is out

**Comment by thaJeztah** [comment]
Ah, yes I was updating, but the images on Docker Hub weren’t available yet when I was.

We also still have to address the failing tests for 1.10 /cc @tonistiigi @dmcgowan

**Comment by thaJeztah** [comment]
Bumped to Go 1.10

**Comment by thaJeztah** [comment]
FWIW, containerd is dropping support for Go 1.9; https://github.com/containerd/cri-containerd/issues/626

**Comment by dmcgowan** [comment]
I tracked down one of the failures to the tarsum code. I will get that resolved and push an update+rebase here. Unless @thaJeztah you would rather close this PR and have me open a new one. I like you owning it though :stuck_out_tongue:

**Comment by dmcgowan** [comment]
One more failure related to go 1.10. I haven't root caused this but was able to determine the panic will not occur with go 1.9. @cpuguy83 do you have any ideas?

```
18:00:39 --- FAIL: TestAdapterReadLogs (0.00s)
18:00:39 panic: runtime error: invalid memory address or nil pointer dereference [recovered]
18:00:39 	panic: runtime error: invalid memory address or nil pointer dereference
18:00:39 [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x761656]
18:00:39 
18:00:39 goroutine 7 [running]:
18:00:39 testing.tRunner.func1(0xc42014c0f0)
18:00:39 	/usr/local/go/src/testing/testing.go:742 +0x29d
18:00:39 panic(0x7d36e0, 0xa7f6a0)
18:00:39 	/usr/local/go/src/runtime/panic.go:502 +0x229
18:00:39 github.com/docker/docker/daemon/logger.testMessageEqual(0xc42014c0f0, 0xc420053a08, 0x0)
18:00:39 	/go/src/github.com/docker/docker/daemon/logger/adapter_test.go:178 +0x56
18:00:39 github.com/docker/docker/daemon/logger.TestAdapterReadLogs(0xc42014c0f0)
18:00:39 	/go/src/github.com/docker/docker/daemon/logger/adapter_test.go:131 +0x895
18:00:39 testing.tRunner(0xc42014c0f0, 0x8633a0)
18:00:39 	/usr/local/go/src/testing/testing.go:777 +0xd0
18:00:39 created by testing.(*T).Run
18:00:39 	/usr/local/go/src/testing/testing.go:824 +0x2e0
18:00:39 FAIL	github.com/docker/docker/daemon/logger	0.005s
```

**Comment by dmcgowan** [comment]
The panic does occur in 1.9 as well, but it is extremely rare,  about 1 in 30,000 `go test -count=30000 -run TestAdapterReadLogs -v ./daemon/logger`. It appears to be a race between writing and reading from the file in the mock implementation.

**Comment by thaJeztah** [comment]
Failures look legit; could it be that for some reason it's picking up the name of the Dockerfile as part of the Dockerfile content?

```
23:07:45 === RUN   TestBuildMultiStageParentConfig
23:07:45 --- FAIL: TestBuildMultiStageParentConfig (0.08s)
23:07:45 	build_test.go:163: assertion failed: error is not nil: Error response from daemon: Dockerfile parse error line 1: unknown instruction: DOCKERFILE0100644000000000000000000000021213
23:07:45 === RUN   TestBuildWithEmptyLayers
23:07:46 --- FAIL: TestBuildWithEmptyLayers (0.08s)
23:07:46 	build_test.go:197: assertion failed: error is not nil: Error response from daemon: Dockerfile parse error line 1: unknown instruction: ASDF
23:07:46 === RUN   TestBuildMultiStageOnBuild
23:07:46 --- FAIL: TestBuildMultiStageOnBuild (0.09s)
23:07:46 	build_test.go:233: assertion failed: error is not nil: Error response from daemon: Dockerfile parse error line 1: unknown instruction: DOCKERFILE0100644000000000000000000000024413260533701013047
````

**Comment by thaJeztah** [comment]
Go 1.10.1 was also released, so looks like we need to bump the version; https://github.com/golang/go/issues?utf8=✓&q=milestone%3AGo1.10.1

**Comment by dmcgowan** [comment]
I am seeing a similar issue in the tests locally but not sure the solution. It seems to be caused by these warnings...
```
18:46:21 # github.com/docker/docker/cmd/dockerd
18:46:21 /tmp/go-link-534967620/000008.o: In function `mygetgrouplist':
18:46:21 /usr/local/go/src/os/user/getgrouplist_unix.go:15: warning: Using 'getgrouplist' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
18:46:21 /tmp/go-link-534967620/000007.o: In function `mygetgrgid_r':
18:46:21 /usr/local/go/src/os/user/cgo_lookup_unix.go:38: warning: Using 'getgrgid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
18:46:21 /tmp/go-link-534967620/000007.o: In function `mygetgrnam_r':
18:46:21 /usr/local/go/src/os/user/cgo_lookup_unix.go:43: warning: Using 'getgrnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
18:46:21 /tmp/go-link-534967620/000007.o: In function `mygetpwnam_r':
18:46:21 /usr/local/go/src/os/user/cgo_lookup_unix.go:33: warning: Using 'getpwnam_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
18:46:21 /tmp/go-link-534967620/000007.o: In function `mygetpwuid_r':
18:46:21 /usr/local/go/src/os/user/cgo_lookup_unix.go:28: warning: Using 'getpwuid_r' in statically linked applications requires at runtime the shared libraries from the glibc version used for linking
```
When one of these functions is called, the syscall is segfaulting. @stevvooe said he has seen this in previous go versions as well, I am not sure if that means it is a musl issue or what the cause may be. Any ideas how to rectify this?

**Comment by justincormack** [comment]
@dmcgowan static builds should either be done in an Alpine container or use `CGO_ENABLED=0`, otherwise they are not static. Or we remove the code using libc for group listings and do it directly.

**Comment by kolyshkin** [comment]
> When one of these functions is called, the syscall is segfaulting

This is not a syscall but a glibc function, otherwise it's legit (i.e. it is not a bug but just the way how things work).

Possible solutions are:
 1. use custom-compiled glibc;
 2. use musl instead of glibc (e.g. by building on Alpine);
 3. do not use `os/user` package (that also includes all dependencies);
 4. wait for https://go-review.googlesource.com/c/go/+/92456 to be merged and use osusergo build tag.

I am currently pursuing 4; in the meantime _some_ packages have workarounds. A few examples from the moby codebase:
 - https://github.com/moby/moby/blob/master/daemon/logger/gcplogs/gcplogging_linux.go
 - https://github.com/moby/moby/blob/master/pkg/homedir/homedir_linux.go
 - https://github.com/moby/moby/blob/master/vendor/github.com/godbus/dbus/homedir_static.go
  - https://github.com/moby/moby/blob/8a9e1808cfdef414e60307b4ccd96719731be304/vendor/golang.org/x/oauth2/google/sdk.go#L162

Now, in case of a segfault the current course of action is to figure out what causes it, and implement a workaround similar to the one above.

**Comment by kolyshkin** [comment]
Failure on experimental ([from here](https://jenkins.dockerproject.org/job/Docker-PRs-experimental/40149/console)):
```
=== RUN   TestAuthZPluginEnsureLoadImportWorking
--- FAIL: TestAuthZPluginEnsureLoadImportWorking (1.06s)
        daemon.go:283: [db66ecf55b2d0] waiting for daemon to start
        daemon.go:315: [db66ecf55b2d0] daemon started
        authz_plugin_test.go:373: assertion failed: error is not nil: Cannot connect to the Docker daemon at unix:///tmp/docker-integration/db66ecf55b2d0.sock. Is the docker daemon running?
        daemon.go:391: Daemon db66ecf55b2d0 is not started
```

is caused by `archive/tar` calling 

```
time="2018-04-05T18:48:58.409988506Z" level=debug msg="AuthZ response using plugin authzplugin"
time="2018-04-05T18:48:58.410703603Z" level=debug msg="Calling GET /v1.37/containers/24099a2641710ca352275c39ecfa51715
ba7e2a3ef4b34afba9fc038c49faa6e/export"
time="2018-04-05T18:48:58.410724822Z" level=debug msg="AuthZ request using plugin authzplugin"
fatal error: unexpected signal during runtime execution
[signal SIGSEGV: segmentation violation code=0x1 addr=0xe5 pc=0x7fa8342d77f8]

runtime stack:
runtime.throw(0x183d93d, 0x2a)
        /usr/local/go/src/runtime/panic.go:616 +0x83
runtime.sigpanic()
        /usr/local/go/src/runtime/signal_unix.go:372 +0x292

goroutine 192 [syscall]:
runtime.cgocall(0x1748570, 0xc42095b5a8, 0x1839600)
        /usr/local/go/src/runtime/cgocall.go:128 +0x66 fp=0xc42095b560 sp=0xc42095b528 pc=0x402236
os/user._Cfunc_mygetpwuid_r(0x3, 0xc4207f94a0, 0x34734c0, 0x400, 0xc4200c4578, 0x0)
        _cgo_gotypes.go:170 +0x4f fp=0xc42095b5a8 sp=0xc42095b560 pc=0x97257f
os/user.lookupUnixUid.func1.1(0x7fa800000003, 0xc4207f94a0, 0x34734c0, 0x400, 0xc4200c4578, 0xc42095b638)
        /usr/local/go/src/os/user/cgo_lookup_unix.go:100 +0x141 fp=0xc42095b5e8 sp=0xc42095b5a8 pc=0x9741c1
os/user.lookupUnixUid.func1(0x10)
        /usr/local/go/src/os/user/cgo_lookup_unix.go:100 +0x52 fp=0xc42095b628 sp=0xc42095b5e8 pc=0x974262
os/user.retryWithBuffer(0xc420b01c60, 0xc42095b718, 0xc420b01c60, 0x2199980)
        /usr/local/go/src/os/user/cgo_lookup_unix.go:253 +0x3d fp=0xc42095b688 sp=0xc42095b628 pc=0x97366d
os/user.lookupUnixUid(0x3, 0x0, 0x0, 0x0)
        /usr/local/go/src/os/user/cgo_lookup_unix.go:96 +0x132 fp=0xc42095b750 sp=0xc42095b688 pc=0x972ac2
os/user.lookupUserId(0x187f9c0, 0x1, 0x1, 0x47df00, 0x3)
        /usr/local/go/src/os/user/cgo_lookup_unix.go:86 +0x75 fp=0xc42095b788 sp=0xc42095b750 pc=0x972955
os/user.LookupId(0x187f9c0, 0x1, 0x1, 0x0, 0x0)
        /usr/local/go/src/os/user/lookup.go:41 +0x53 fp=0xc42095b7d8 sp=0xc42095b788 pc=0x971f23
archive/tar.statUnix(0x230c640, 0xc4213e45b0, 0xc420ede2a0, 0x17ecf6e, 0x1)
        /usr/local/go/src/archive/tar/stat_unix.go:39 +0x4c3 fp=0xc42095b858 sp=0xc42095b7d8 pc=0x97dd33
archive/tar.FileInfoHeader(0x230c640, 0xc4213e45b0, 0x0, 0x0, 0xc420559360, 0x99, 0x230c640)
        /usr/local/go/src/archive/tar/common.go:699 +0x493 fp=0xc42095b9d8 sp=0xc42095b858 pc=0x9771b3
github.com/docker/docker/pkg/archive.FileInfoHeader(0xc4205593f6, 0x3, 0x230c640, 0xc4213e45b0, 0x0, 0x0, 0x7412596f, 0xf943ee04c5fd7027, 0x0)
        /go/src/github.com/docker/docker/pkg/archive/archive.go:360 +0x55 fp=0xc42095ba38 sp=0xc42095b9d8 pc=0xd73705
github.com/docker/docker/pkg/archive.(*tarAppender).addTarFile(0xc4211b70c0, 0xc420559360, 0x99, 0xc4205593f6, 0x3, 0x0, 0x0)
        /go/src/github.com/docker/docker/pkg/archive/archive.go:478 +0xd9 fp=0xc42095bad0 sp=0xc42095ba38 pc=0xd73f99
github.com/docker/docker/pkg/archive.TarWithOptions.func1.2(0xc420559360, 0x99, 0x230c640, 0xc4213e44e0, 0x0, 0x0, 0x0, 0xc42095bd28)
        /go/src/github.com/docker/docker/pkg/archive/archive.go:888 +0x612 fp=0xc42095bca0 sp=0xc42095bad0 pc=0xd7ed42
path/filepath.walk(0xc420559360, 0x99, 0x230c640, 0xc4213e44e0, 0xc4200c74f0, 0x0, 0x0)
        /usr/local/go/src/path/filepath/path.go:361 +0xe7 fp=0xc42095bd78 sp=0xc42095bca0 pc=0x50fd17
path/filepath.walk(0xc4213e79a0, 0x97, 0x230c640, 0xc420eac8f0, 0xc4200c74f0, 0x0, 0x50)
        /usr/local/go/src/path/filepath/path.go:381 +0x2c4 fp=0xc42095be50 sp=0xc42095bd78 pc=0x50fef4
path/filepath.Walk(0xc4213e79a0, 0x97, 0xc4200c74f0, 0x17ecf6e, 0x1)
        /usr/local/go/src/path/filepath/path.go:403 +0x108 fp=0xc42095beb0 sp=0xc42095be50 pc=0x510178
github.com/docker/docker/pkg/archive.TarWithOptions.func1(0xc4202a25a0, 0x22eb160, 0xc4203cbec0, 0xc4200c5e48, 0xc4200abfb0, 0xc4203cbea0)
        /go/src/github.com/docker/docker/pkg/archive/archive.go:806 +0x2ee fp=0xc42095bfb0 sp=0xc42095beb0 pc=0xd7f42e
runtime.goexit()
        /usr/local/go/src/runtime/asm_amd64.s:2361 +0x1 fp=0xc42095bfb8 sp=0xc42095bfb0 pc=0x45bc71
created by github.com/docker/docker/pkg/archive.TarWithOptions
        /go/src/github.com/docker/docker/pkg/archive/archive.go:748 +0x26f
```

**Comment by kolyshkin** [comment]
The above is caused by archive/tar (updated in Go 1.10) which now does user/group lookups as per earlier TODO in the code (see commit https://github.com/golang/go/commit/0564e304a6ea394a42929060c588469dbd6f32af).

**Comment by kolyshkin** [comment]
OK I see a few ways to fix this:
1. implement own `tar.FileInfoHeader()` for the case of `static_build` (essentially copy/pasting the code from Go 1.9 implementation of it (before https://github.com/golang/go/commit/0564e304a6ea394a42929060c588469dbd6f32af)).
2a. backport https://go-review.googlesource.com/c/go/+/92456 (by patching Go right there in /usr/local/go) and use `osusergo` build tag for static build.
2b. have https://go-review.googlesource.com/c/go/+/92456 officially backported to Go 1.10
3a. revert https://github.com/golang/go/commit/0564e304a6ea394a42929060c588469dbd6f32af (by patching Go right there in /usr/local/go).
3b. revert https://github.com/golang/go/commit/0564e304a6ea394a42929060c588469dbd6f32af (by re-vendoring archive/tar as we used to do)
4. switch to Alpine.

**Comment by kolyshkin** [comment]
I have asked about backporting https://go-review.googlesource.com/c/go/+/92456 to Go 1.10 (as per 2b [above](https://github.com/moby/moby/pull/35739#issuecomment-379405517) and got a no.

Going to proceed with 1 (implement own `tar.FileInfoHeader()`)

**Comment by kolyshkin** [comment]
btw containerd has the same problem when compiled statically

**Comment by kolyshkin** [comment]
Filed https://github.com/golang/go/issues/24787 just for the sake of it, let's see how it goes.

Also created https://github.com/kolyshkin/go-tar/tree/go-1.10 as a workaround, will make a patch tomorrow

**Comment by kolyshkin** [comment]
windows ci failure:

> 18:01:48 These files import internal code: (either directly or indirectly)
> 18:01:48  - pkg/tarsum/tarsum_test.go imports False

not sure where is this coming from :-\ but it might be related to the latest commit

**Comment by thaJeztah** [comment]
Error comes from https://github.com/moby/moby/blob/a1be987ea9e03e5ebdb1b415a7acdd8d6f0aaa08/hack/make.ps1#L247-L266 (on linux: https://github.com/moby/moby/blob/4163b5cbd59e2a5dee411310cb49734472f95440/hack/validate/pkg-imports#L12-L33)

**Comment by kolyshkin** [comment]
> Error comes from ...

@johnstep can you PTAL? I have zero powershell knowledge :(

**Comment by johnstep** [comment]
In Go 1.10.1, the `Deps` for `pkg\tarsum\tarsum_test.go` are empty (`[]`) and the PowerShell script ends up setting its import list to a string value of `False` instead of an empty array. This can be remedied by forcing the result to be an array, by concatenating to an empty array (`@() + ...`):

```diff
--- a/hack/make.ps1
+++ b/hack/make.ps1
@@ -251,10 +251,10 @@ Function Validate-PkgImports($headCommit, $upstreamCommit) {
         if ($LASTEXITCODE -ne 0) { Throw "Failed go list for dependencies on $file" }
         $imports = $imports -Replace "\[" -Replace "\]", "" -Split(" ") | Sort-Object | Get-Unique
         # Filter out what we are looking for
-        $imports = $imports -NotMatch "^github.com/docker/docker/pkg/" `
-                            -NotMatch "^github.com/docker/docker/vendor" `
-                            -Match "^github.com/docker/docker" `
-                            -Replace "`n", ""
+        $imports = @() + $imports -NotMatch "^github.com/docker/docker/pkg/" `
+                                  -NotMatch "^github.com/docker/docker/vendor" `
+                                  -Match "^github.com/docker/docker" `
+                                  -Replace "`n", ""
         $imports | % { $badFiles+="$file imports $_`n" }
     }
```

However, I was curious why the list is empty. It seems that, for test files, the imported packages are listed in `TestImports` instead of `Deps`:
```PowerShell
$ go list -e -f "{{ .Deps }}`n{{ .TestImports }}" tarsum_test.go
[]
[bytes compress/gzip crypto/md5 crypto/rand crypto/sha1 crypto/sha256 crypto/sha512 encoding/hex fmt github.com\moby\moby/vendor/archive/tar github.com\moby\moby/vendor/github.com/gotestyourself/gotestyourself/assert github.com\moby\moby/vendor/github.com/gotestyourself/gotestyourself/assert/cmp io io/ioutil os strings testing]
```

Should we be verifying `TestImports` as well?

**Comment by kolyshkin** [comment]
Current CI failures are

### experimental:
> A20:05:58 FAIL: check_test.go:362: DockerSwarmSuite.TearDownTest

This looks like a subject of https://github.com/moby/moby/issues/35825

### janky

> 21:07:15 Step 1/28 : FROM golang:1.10.1-alpine3.6 as builder
> 21:07:16 manifest for golang:1.10.1-alpine3.6 not found
> 21:07:16 Build step 'Execute shell' marked build as failure

A temp network glitch?

### z

Some glitch at the very start of the CI cycle; restarted

**Comment by kolyshkin** [comment]
> 12:31:14 Step 1/28 : FROM golang:1.10.1-alpine3.6 as builder
> 12:31:15 manifest for golang:1.10.1-alpine3.6 not found

OK this is a legit error, as there's no such image, the closest one is golang:1.10.1-alpine3.7. I have a patch ready at https://github.com/kolyshkin/moby/tree/go-1.10-fix-tar

**Comment by kolyshkin** [comment]
z failure might be a flaky test (there is https://github.com/moby/moby/issues/32673 but it looks different)
```
11:04:36 ----------------------------------------------------------------------
11:04:36 FAIL: docker_api_swarm_test.go:297: DockerSwarmSuite.TestAPISwarmLeaderElection
11:04:36 
11:04:36 [dfd51c688eb05] waiting for daemon to start
11:04:36 [dfd51c688eb05] daemon started
11:04:36 
11:04:36 [db54885b03907] waiting for daemon to start
11:04:36 [db54885b03907] daemon started
11:04:36 
11:04:36 [da336d294b579] waiting for daemon to start
11:04:36 [da336d294b579] daemon started
11:04:36 
11:04:36 [dfd51c688eb05] exiting daemon
11:04:36 assertion failed: error is not nil: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded
11:04:36 [db54885b03907] exiting daemon
11:04:36 [da336d294b579] exiting daemon
11:05:00 
11:05:00 ----------------------------------------------------------------------
```

**Comment by kolyshkin** [comment]
Good news we're down to a single test failure on experimental. Bad news it looks it is consistent. I have restarted it to see if that is the case. Here's the failure:

```
14:43:53 ----------------------------------------------------------------------
14:43:53 FAIL: docker_api_swarm_test.go:297: DockerSwarmSuite.TestAPISwarmLeaderElection
14:43:53 
14:43:53 [d60173fcc715d] waiting for daemon to start
14:43:53 [d60173fcc715d] daemon started
14:43:53 
14:43:53 [d9b8e91073dd1] waiting for daemon to start
14:43:53 [d9b8e91073dd1] daemon started
14:43:53 
14:43:53 [d96d4e5a515e0] waiting for daemon to start
14:43:53 [d96d4e5a515e0] daemon started
14:43:53 
14:43:53 [d60173fcc715d] exiting daemon
14:43:53 assertion failed: error is not nil: Error response from daemon: rpc error: code = DeadlineExceeded desc = context deadline exceeded
14:43:53 [d9b8e91073dd1] exiting daemon
14:43:53 [d96d4e5a515e0] exiting daemon
```

**Comment by thaJeztah** [comment]
I think I've seen that test fail recently; it may not be related, but definitely something to look into

**Comment by thaJeztah** [comment]
It's greeeeeeeeeeen 🎉

**Comment by tonistiigi** [comment]
LGTM

**Comment by tiborvass** [comment]
LGTM

**Comment by thaJeztah** [comment]
Yay! Thanks all for helping out on this one!

**Pull request merged**